### ISSUE TITLE
Regression: PEP 695 type alias handling causes syntax errors in Python <3.12 environments (PR #10488)

### ISSUE THREAD

**Issue by wenwen-666** [PyCQA/pylint#10907]
### Bug description

```python
PR #10488 ("Fix used-before-assignment for PEP 695 type aliases and parameters") added a new utility function is_node_in_pep695_type_context that parses PEP 695 syntax without version guards, causing AstroidSyntaxError when running in Python <3.12 environments.
Regression Evidence
My automated probe tests that directly test the new function show:
Test code (aug_isnodeinpep695typecontext_8):
import astroid
from pylint.checkers.utils import is_node_in_pep695_type_context

print('=== Probe for New Path [0] ===')
print('Testing node with TypeAlias ancestor returns that TypeAlias node.')
print('Step 1: Parse code with TypeAlias (PEP 695).')
source = '''
type MyAlias = int
'''
module = astroid.parse(source)
type_alias_node = module.body[0]
value_node = type_alias_node.value
result = is_node_in_pep695_type_context(value_node)
Output BEFORE PR #10488:
exception_type: "ImportError"
exception_message: "cannot import name 'is_node_in_pep695_type_context' from 'pylint.checkers.utils'"
Output AFTER PR #10488:
exception_type: "AstroidSyntaxError"
exception_message: "Parsing Python code failed: invalid syntax (<unknown>, line 2)"
stdout: "=== Probe for New Path [0] ===
Testing node with TypeAlias ancestor returns that TypeAlias node.
Step 1: Parse code with TypeAlias (PEP 695)."
Test code (aug_isnodeinpep695typecontext_9):
import astroid
from pylint.checkers.utils import is_node_in_pep695_type_context

print('=== Probe for New Path [1] ===')
print('Testing node with TypeVar ancestor returns that TypeVar node.')
print('Step 1: Parse code with generic function using TypeVar (PEP 695).')
source = '''
def func[T](x: T) -> T:
    return x
'''
module = astroid.parse(source)
func_node = module.body[0]
type_var_node = func_node.type_params[0]
result = is_node_in_pep695_type_context(type_var_node)
Output BEFORE PR #10488:
exception_type: "ImportError"
exception_message: "cannot import name 'is_node_in_pep695_type_context' from 'pylint.checkers.utils'"
Output AFTER PR #10488:
exception_type: "AstroidSyntaxError"
exception_message: "Parsing Python code failed: invalid syntax (<unknown>, line 2)"
stdout: "=== Probe for New Path [1] ===
Testing node with TypeVar ancestor returns that TypeVar node.
Step 1: Parse code with generic function using TypeVar (PEP 695)."
Root Cause Analysis
The PR added is_node_in_pep695_type_context to pylint/checkers/utils.py:
def is_node_in_pep695_type_context(node: nodes.NodeNG) -> nodes.NodeNG | None:
    """Check if node is used in a TypeAlias or as part of a type param."""
    return get_node_first_ancestor_of_type(
        node, (nodes.TypeAlias, nodes.TypeVar, nodes.ParamSpec, nodes.TypeVarTuple)
    )
The problem: This function relies on AST nodes (TypeAlias, TypeVar, etc.) that are only available when parsing Python 3.12+ code. When running in Python <3.12 environments, astroid.parse() fails on PEP 695 syntax, causing AstroidSyntaxError.
Why This Is a Regression
Version Compatibility Broken: The PR introduced code that assumes PEP 695 syntax without checking Python version

Silent Failure: Users running pylint on Python 3.8-3.11 will encounter syntax errors when analyzing code that uses PEP 695 features

No Version Guard: The new function is called unconditionally in variables.py, even on older Python versions:

if (
    self._postponed_evaluation_enabled
    and utils.is_node_in_type_annotation_context(node)
) or utils.is_node_in_pep695_type_context(node):  # Called unconditionally
    return False
```

### Configuration

```ini
Any configuration that analyzes Python 3.12+ code with PEP 695 features, when running on Python <3.12.
```

### Command used

```shell
# Running on Python 3.10, analyzing Python 3.12 code with PEP 695 syntax
pylint myfile.py  # myfile.py contains 'type MyAlias = int' or generic functions
```

### Pylint output

```python
AstroidSyntaxError: Parsing Python code failed:
invalid syntax (<unknown>, line 2)
```

### Expected behavior

Pylint should gracefully handle PEP 695 syntax when running on older Python versions, either by:

Skipping the new checks when Python version <3.12

Using version guards to avoid parsing PEP 695 syntax on incompatible versions

Falling back to a no-op implementation for older Python versions

### Pylint version

```shell
Tested on:

Before PR #10488: pylint-dev/pylint@main (prior to merge) - function doesn't exist

After PR #10488: pylint-dev/pylint@main (after merge, commit cdce8p) - function exists but causes syntax errors on Python <3.12

Affected versions: All versions after PR #10488 was merged (August 15, 2025), when running on Python <3.12
pylint 3.3.0 (and later)
astroid 3.3.0
Python 3.10.0 (test environment)
```

### OS / Environment

Any environment running Python <3.12

### Additional dependencies

```python

```

**Comment by DanielNoord** [comment]
No. If you're using new syntax you should also use a Python version and a `pylint` version that supports that syntax. This has always been the case and won't change.

**Comment by Pierre-Sassoulas** [comment]
Because pylint rely on the builtin ast and the python interpreter to parse the code we can handle older syntax from newer interpreter but not newer syntax from older interpreter. This is a fundamental of pylint, it's easier to start a new tool from scratch than challenge this assumption.
